### PR TITLE
PHPC-181: Const is our friend

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-181

### CODE REVIEW

File: php_phongo.c [hunk 1]
```diff
@@ -1559,7 +1559,7 @@ void _phongo_debug_bson(bson_t *bson)
 
 /* {{{ INI entries */
 PHP_INI_BEGIN()
-	STD_PHP_INI_ENTRY("phongo.debug_log", (char *)"", PHP_INI_ALL, OnUpdateString, debug_log, zend_phongo_globals, phongo_globals)
+	STD_PHP_INI_ENTRY((char *)"phongo.debug_log", (char *)"", PHP_INI_ALL, OnUpdateString, debug_log, zend_phongo_globals, phongo_globals)
```
Comment (jmikola) [inline php_phongo.c]: I think you missed this when adding `(char *)""`.

File: php_phongo.c [hunk 2]
```diff
@@ -1066,9 +1066,9 @@ void php_phongo_write_concern_to_zval(zval *retval, mongoc_write_concern_t *writ
 	array_init_size(retval, 5);
 
 	if (wtag) {
-		add_assoc_string_ex(retval, ZEND_STRS("w"), wtag, 1);
+		add_assoc_string_ex(retval, ZEND_STRS("w"), (char *)wtag, 1);
 	} else if (mongoc_write_concern_get_wmajority(write_concern)) {
-		add_assoc_string_ex(retval, ZEND_STRS("w"), "majority", 1);
+		add_assoc_string_ex(retval, ZEND_STRS("w"), (char *)"majority", 1);
```
Comment (jmikola) [inline php_phongo.c]: This should be fine provided we don't compile with `-Wcast-qual`.

File: php_phongo.h [hunk 3]
```diff
@@ -105,22 +105,22 @@ PHONGO_API zend_object_handlers *phongo_get_std_object_handlers(void);
...
 
 mongoc_stream_t*         phongo_stream_initiator     (const mongoc_uri_t *uri, const mongoc_host_list_t *host, void *user_data, bson_error_t *error);
 zend_object_iterator*    phongo_result_get_iterator  (zend_class_entry *ce, zval *object, int by_ref TSRMLS_DC);
 zend_object_iterator*    phongo_cursor_get_iterator  (zend_class_entry *ce, zval *object, int by_ref TSRMLS_DC);
-mongoc_read_prefs_t*     phongo_read_preference_from_zval(zval *zread_preference TSRMLS_DC);
-mongoc_write_concern_t*  phongo_write_concern_from_zval  (zval *zwrite_concern TSRMLS_DC);
-php_phongo_query_t*      phongo_query_from_zval          (zval *zquery TSRMLS_DC);
+const mongoc_read_prefs_t*    phongo_read_preference_from_zval(zval *zread_preference TSRMLS_DC);
+const mongoc_write_concern_t* phongo_write_concern_from_zval  (zval *zwrite_concern TSRMLS_DC);
+const php_phongo_query_t*     phongo_query_from_zval          (zval *zquery TSRMLS_DC);
```
Comment (jmikola) [inline php_phongo.h]: If these should be realigned, I can do so in a subsequent commit.

File: src/MongoDB/ReadPreference.c [hunk 4]
```diff
@@ -147,7 +147,7 @@ zend_object_value php_phongo_readpreference_create_object(zend_class_entry *clas
 HashTable *php_phongo_readpreference_get_debug_info(zval *object, int *is_temp TSRMLS_DC) /* {{{ */
 {
 	zval                 retval = zval_used_for_init;
-	mongoc_read_prefs_t *read_prefs = phongo_read_preference_from_zval(object TSRMLS_CC);
+	const mongoc_read_prefs_t *read_prefs = phongo_read_preference_from_zval(object TSRMLS_CC);
```
Comment (jmikola) [inline src/MongoDB/ReadPreference.c]: I originally thought of getting rid of this declaration and just passing the return value directly below. I then realized that `php_phongo_read_preference_to_zval()` doesn't actually check that this is a valid pointer, and `phongo_read_preference_from_zval()` may return `NULL`. Should we add a sanity check here?

File: php_phongo.h (hunk 3 above)
Comment (bjori) [inline php_phongo.h]: oh sounds good. if you are bored, feel free to realign and organize the entire file to be pretty!
I've been meaning to do that for a while

File: src/MongoDB/ReadPreference.c (hunk 4 above)
Comment (bjori) [inline src/MongoDB/ReadPreference.c]: Is there a theoretical way it can return NULL?

**Comment by bjori** [comment]
This passes with --enable-developer-flags I imagine?
If so, lgtm

**Comment by jmikola** [comment]
Only errors with `--enable-developer-flags` is:

```
In file included from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-compat.h:28:0,
                 from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson.h:23,
                 from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-iter.h:27,
                 from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-iter.c:18:
/home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-iter.c: In function ‘bson_iter_overwrite_bool’:
/home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-iter.c:1956:36: warning: logical ‘or’ of collectively exhaustive tests is always true [-Wlogical-op]
    bson_return_if_fail (value == 1 || value == 0);
                                    ^
/home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-macros.h:204:13: note: in definition of macro ‘bson_return_if_fail’
       if (!(test)) { \
             ^
In file included from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-compat.h:28:0,
                 from /home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson.h:23,
                 from /home/jmikola/workspace/mongodb/phongo/src/libmongoc/src/mongoc/mongoc-buffer.c:18:
/home/jmikola/workspace/mongodb/phongo/src/libmongoc/src/mongoc/mongoc-buffer.c: In function ‘_mongoc_buffer_fill’:
/home/jmikola/workspace/mongodb/phongo/src/libmongoc/src/mongoc/mongoc-buffer.c:216:38: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
    bson_return_val_if_fail(min_bytes >= 0, false);
                                      ^
/home/jmikola/workspace/mongodb/phongo/src/libbson/src/bson/bson-macros.h:218:13: note: in definition of macro ‘bson_return_val_if_fail’
       if (!(test)) { \
             ^
/home/jmikola/workspace/mongodb/phongo/php_phongo.c: In function ‘phongo_log_writer’:
/home/jmikola/workspace/mongodb/phongo/php_phongo.c:232:28: warning: unused variable ‘base_stream’ [-Wunused-variable]
  php_phongo_stream_socket *base_stream = (php_phongo_stream_socket *)stream;
                            ^
```

libbson is what it is, and the other warning is only because your TODO isn't implemented.

File: src/MongoDB/ReadPreference.c (hunk 4 above)
Comment (jmikola) [inline src/MongoDB/ReadPreference.c]: Anything is possible.

File: php_phongo.h (hunk 3 above)
Comment (jmikola) [inline php_phongo.h]: Let's save that before the final release in case we introduce more changes.

**Pull request merged**